src/AnalysisEpic.cxx

```diff
@@ -0,0 +1,424 @@
+    std::map <double,int> genidmap; // <pz, index_gen>
+    std::map <int,int> mcidmap; // <index_mc, index_gen>
+    std::map <int,int> trackidmap; // <index, index_mc>  
+    std::map <int,int> trackstatmap; // <index, genstatus>
+
+    // Particles vectors
+    // The index of the vectors correspond to their for loop idx
+    std::vector<Particles> genpart;    // mcID --> igen
+    std::vector<Particles> mcpart;     // mcID --> imc
+    std::vector<Particles> trackpart;  // mcID --> (imc of matching mcpart) or (-1 if no match is found)
+
+    /*
+      GenParticles loop
+    */
+
+    for(int igen=0; igen<hepmcp_PDG.GetSize(); igen++) {
+
+      int pid_ = hepmcp_PDG[igen];
+
+      double px_ = hepmcp_psx[igen];
+      double py_ = hepmcp_psy[igen];
+      double pz_ = hepmcp_psz[igen];
+      double e_  = hepmcp_E[igen];
+     
+      double p_ = sqrt(pow(hepmcp_psx[igen],2) + pow(hepmcp_psy[igen],2) + pow(hepmcp_psz[igen],2));
+      double mass_ = (fabs(pid_)==211)?constants::pimass:(fabs(pid_)==321)?constants::kmass:(fabs(pid_)==11)?constants::emass:(fabs(pid_)==13)?constants::mumass:(fabs(pid_)==2212)?constants::pmass:0.;
+
+      // Add to genpart
+      Particles part;
+
+      part.pid=pid_;
+      part.charge = (pid_ == 211 || pid_ == 321 || pid_ == 2212 || pid_ == -11 || pid_ == -13)?1:(pid_ == -211 || pid_ == -321 || pid_ == -2212 || pid_ == 11 || pid_ == 13)?-1:0;
+      part.mcID=igen;
+      part.vecPart.SetPxPyPzE(px_,py_,pz_,e_);
+      genpart.push_back(part);
+      
+      genidmap.insert({pz_,igen});
+
+    }
+
+    /*
+      MCParticles loop
+    */
+
+    for(int imc=0; imc < mcpart_PDG.GetSize(); imc++){
+
+      int pid_ = mcpart_PDG[imc];
+      double px_ = mcpart_psx[imc];
+      double py_ = mcpart_psy[imc];
+      double pz_ = mcpart_psz[imc];
+      double m_ = mcpart_m[imc];
+      double e_ = sqrt(px_*px_+py_*py_+pz_*pz_+m_*m_);
+
+      // Add to mcpart
+      Particles part;
+
+      part.pid=pid_;
+      part.charge = (pid_ == 211 || pid_ == 321 || pid_ == 2212 || pid_ == -11 || pid_ == -13)?1:(pid_ == -211 || pid_ == -321 || pid_ == -2212 || pid_ == 11 || pid_ == 13)?-1:0;
+      part.mcID=imc;
+      part.vecPart.SetPxPyPzE(px_,py_,pz_,e_);
+      mcpart.push_back(part);
+      
+      int igen=-1;
+      if(auto search = genidmap.find(pz_); search != genidmap.end())
+	igen=search->second; //index of the GeneratedParticle
+      
+      mcidmap.insert({imc,igen});
+
+    }
+
+    /*
+      ReconstructedParticles loop
+      - Add all particles to the std::vector<> of particles
+      - Identify the 
+      - Identify closest matching MCParticle in theta,phi,E space
+
+    */
+
+   
+      
+    for(int itrack=0; itrack < tracks_PDG.GetSize(); itrack++){
+
+      int pid_ = tracks_PDG[itrack];
+      double px_ = tracks_p_x[itrack];
+      double py_ = tracks_p_y[itrack];
+      double pz_ = tracks_p_z[itrack];
+      double e_ = tracks_e[itrack];
+      double m_ = sqrt(e_*e_-px_*px_+py_*py_+pz_*pz_);
```

Comment:
is this formula correct? You use these to make a TLorentzVector below, why not let TLorentzVector calculate the mass for you?